test/test_autograd.py

```diff
@@ -4558,6 +4559,63 @@ def parent_on_cpu(inp):
         # gpu thread ReadyQueue
         out.sum().backward()
 
+    @onlyCUDA
+    def test_reentrant_none_blocking_parent(self, device):
+        # Typically parent node will run after child reentrant node. However if the reentrant
+        # node runs on a different device, it's possible for the parent node to run ahead of the
+        # reentrant node.
+        q = queue.Queue()
+        order = []
+
+        class ParentFunc(Function):
+
+            @staticmethod
+            def forward(ctx, inp):
+                return 2 * inp
+
+            @staticmethod
+            def backward(ctx, grad):
+                order.append("Parent")
+                q.put(0)
+                return 2 * grad
+
+        class ChildFunc(Function):
+
+            @staticmethod
+            def forward(ctx, inp):
+                ctx.save_for_backward(inp)
+                return inp * inp
+
+            @staticmethod
+            def backward(ctx, grad):
+                q.get(block=True)
+                order.append("Child")
+                inp, = ctx.saved_tensors
+                return grad * 2 * inp
+
+        a0 = torch.randn(3, 3, requires_grad=True)
+        a1 = ParentFunc.apply(a0)
+
+        re_input = torch.rand(3, 3, requires_grad=True, device=device)
+        re_loss = ChildFunc.apply(re_input).sum()
+
+        class ReentrantFunc(Function):
+
+            @staticmethod
+            def forward(ctx, inp):
+                return inp
```

Comment:
nit: In general I would advise doing `inp.clone()` to avoid any issues in case you do inplace later. But it's fine here if you want to leave it as-is.